### PR TITLE
Improve free-hand drawing of sliders to the editor

### PR DESCRIPTION
**Pull request by OliBomby**
- [x] Depends on https://github.com/ppy/osu-framework/pull/6056
- [x] Depends on https://github.com/ppy/osu-framework/pull/6066

This PR implements the improvements made to `IncrementalBSplineBuilder` and changes some behaviour of the slider drawing to be better in my opinion.

- Change how segments are created from the output of `IncrementalBSplineBuilder`. Because it gives the segments explicitly, the logic becomes a lot simpler.
- Increase the default tolerance value for slider drawing. The improved builder algorithm needs fewer control points to match the input path so a higher tolerance is preferred.
- Remove the behaviour of 'simplifying' linear segments to a merged linear type segment. This seems to be something that peppy added in as an afterthought, but I found that in most cases this makes editing the slider afterwards a lot more painful. In most cases you draw a slider, you have some segments that are either linear when you want them to be curved or curved when you want them to be linear. In those cases its easier if you can just add or remove a control point, but if linear type control points are mixed-in, then you have to do a lot of changing control point path types, which is a pain.
- Prevent snapping slider expected distance while drawing the slider. This makes the drawing a lot smoother and a joy to use.
- Changed default B-Spline order to 4.
- Added Perfect Curve type segments and a threshold parameter.

https://github.com/ppy/osu/assets/17460441/6e8c4187-f745-44b7-9cb6-58d31833c930

https://github.com/ppy/osu/assets/17460441/74380e9d-05f9-4612-abe6-7dfa7e8adfbb

### CODE REVIEW

**Comment by Karoo13** [comment]
3rd order is sufficient to make smooth looking curves, but doing so requires the anchors to be arranged quite smoothly and densely to get a smooth looking path. Adjusting each point of a 3rd order b-spline will require adjusting the surrounding points to maintain smoothness, while 4th order offers some wiggle room to adjust just one anchor without creating ugly curvature spikes. If you need to fiddle around with a large number of anchors in a specific way to refine your shape after drawing, that defeats the purpose of the tool. I disagree that this is a tool for inexperienced mappers, it will be how everyone makes sliders if the shape can be easily refined to a high quality after drawing, and for that it needs to be 4th order or higher.

I did some tests on the ability to create a perfect circular arc using b-splines, and 3rd order you need to roughly place points in an octagon, while 4th order placing points in a pentagon will do. Actual data in the table below. Not all curves being drawn will be circular arcs, but this is a good rule of thumb for the density of anchors needed to create a smooth bend. At the minimum level of precision needed to accurately create a desired shape, 3rd order will take about 1.64 times as many anchors to do so, and as you provide excess anchors (an excess is necessary to allow some flexibility in editing), 4th order will scale better in how much control that adds (as seen by the ratio increasing). The ratios are much higher when going from 3rd to 4th order compared to the surrounding jumps, meaning 4th order is the critical point where creating smooth arcs goes from hard to natural, while going from 2nd to 3rd or 4th to 5th order are just incremental improvements.

`visual error | 2nd order angle | 3rd order angle | 4th order angle | 5th order angle | angle ratio 3:2 | angle ratio 4:3 | angle ratio 5:4`
`0.50% | 52.2 | 68.8 | 94.2 | 1069 | 1.32 | 1.37 | 1.13`
`0.25% | 41.8 | 56.6 | 83.3 | 97.0 | 1.35 | 1.47 | 1.17`
`0.10% | 31.0 | 42.9 | 70.3 | 84.3 | 1.38 | 1.64 | 1.20`
`0.05% | 24.7 | 34.4 | 61.2 | 75.0 | 1.39 | 1.78 | 1.23`
Error .5 is visibly lumpy, .25 is getting hard to tell without a reference circle, .1 is hard to tell with a reference circle, .05 is around where you have enough excess control to comfortably edit the shape. An angle of 60 for example would mean repeatedly placing anchors in a hexagon creates a circle with that error. The definition of visual error is height*frequency of lumps, `(max_r-min_r)/(max_r+min_r)*pi/2/angle`

In summary, I quantified how smooth 2nd vs 3rd vs 4th vs 5th order b-splines are, and the data supports the conclusion that 4th order is the optimal middleground between smoothness and local control. I would advise only 1st (linear), 4th, and infinite (bezier) be available, since giving mappers a 3rd order b-spline to edit the anchors of is just going to make their life harder.

**Comment by peppy** [comment]
I'm also all for 4th order, but would want some feedback from @default0 / @Tom94  😄

**Comment by Tom94** [comment]
No qualms with 4th order at all! 3rd seemed to work better with the old (more naive) control point placement algorithm. And looking at this PR and the discussion, the improved one w/ 4th order looks nicer in pretty much every way.

**Review by peppy — APPROVED** [review]
